server/tests/classes/client.test.js

```diff
@@ -0,0 +1,347 @@
+      expect(k.attempts).toBe(0);
+      expect(k.locked).toBe(false);
+    });
+        
+    test('should set random code of length codeLength', () => {
+      const c = new Client({ keypad: { codeLength: 6 } } );
+      expect(c.keypad.codeLength).toBe(6);
+      expect(c.keypad.code).toBeKeypadCode(6);
+    });
+    
+    test('should set codeLength to match code.length', () => {
+      const c = new Client({ keypad: { code: [1, 2], codeLength: 7 } } );
+      expect(c.keypad.codeLength).toBe(2);
+      expect(c.keypad.code).toBeKeypadCode(2);
+    });
+    
+    test('should restrict code length to 8 digits', () => {
+      const c = new Client({ keypad: { codeLength: 24 } } );
+      expect(c.keypad.codeLength).toBe(8);
+      expect(c.keypad.code).toBeKeypadCode(8);
+
+      const c2 = new Client({ keypad: { code: [1, 2, 3, 4, 5, 6, 7, 8, 9] } } );
+      expect(c.keypad.codeLength).toBe(8);
+      expect(c.keypad.code).toBeKeypadCode(8);
+
+    });
+  })
+
+  describe('setCode', () => {
+    test('should set a specified code', () => {
+      const c = new Client();
+      c.keypad.setCode([1, 2, 3, 4, 5, 6]);
+      expect(c.keypad.code).toEqual([1, 2, 3, 4, 5, 6]);
+    });
+
+    test('should generate a random code', () => {
+      const c = new Client();
+      c.keypad.setCodeLength(7);
+      c.keypad.setCode();
+      expect(c.keypad.code).toBeKeypadCode(7);
+    });
+
+    test('should set codeLength', () => {
+      const c = new Client();
+      c.keypad.setCode([1, 2, 3, 4, 5, 6]);
+      expect(c.keypad.codeLength).toBe(6);
+    });
+
+  });
+
+  describe('setEnteredCode', () => {
+    test('should enter a code, increment attempts, and lock keypad if exceeds maximum attempts', () => {
+      const c = new Client({ keypad: { allowedAttempts: 2 } } );
+      c.keypad.setEnteredCode([1, 2, 3, 4]);
+      expect(c.keypad.enteredCode).toEqual([1, 2, 3, 4]);
+      expect(c.keypad.locked).toBe(false);
+      expect(c.keypad.attempts).toBe(1);
+      c.keypad.setEnteredCode([1, 2, 3, 4]);
+      expect(c.keypad.locked).toBe(true);
+      expect(c.keypad.attempts).toBe(2);
+    });
+  });
+
+  describe('reset', () => {
+    test('should reset the keypad', () => {
+      const c = new Client({ keypad: { allowedAttempts: 2 } } );
+      c.keypad.setEnteredCode([1, 2, 3, 4]);
+      c.keypad.setEnteredCode([1, 2, 3, 4]);
+      expect(c.keypad.enteredCode).toEqual([1, 2, 3, 4]);
+      expect(c.keypad.locked).toBe(true);
+      expect(c.keypad.attempts).toBe(2);
+      c.keypad.reset();
+      expect(c.keypad.enteredCode).toEqual([]);
+      expect(c.keypad.locked).toBe(false);
+      expect(c.keypad.attempts).toBe(0);
+    });
+  });
+
+  describe('setCodeLength', () => {
+    test('should set the code length between 1 and 8 digits', () => {
+      const c = new Client();
+      c.keypad.setCodeLength(5);
+      expect(c.keypad.codeLength).toBe(5);
+
+      // TODO: What if codeLength doesn't match code.length?
```

Comment:
It looks like you adjusted the code so `codeLength` is set to the length of the code you enter, which is the ideal behavior. `codeLength`, on the other hand, is only for randomly generated codes. So I would ignore the length of `this.code` in this test.


```suggestion
```